libs/community/langchain_community/chat_models/edenai.py

```diff
@@ -0,0 +1,337 @@
+import json
+from typing import Any, AsyncIterator, Dict, Iterator, List, Optional
+
+from aiohttp import ClientSession
+from langchain_core.callbacks import (
+    AsyncCallbackManagerForLLMRun,
+    CallbackManagerForLLMRun,
+)
+from langchain_core.language_models.chat_models import (
+    BaseChatModel,
+    agenerate_from_stream,
+    generate_from_stream,
+)
+from langchain_core.messages import (
+    AIMessage,
+    AIMessageChunk,
+    BaseMessage,
+)
+from langchain_core.outputs import ChatGeneration, ChatGenerationChunk, ChatResult
+from langchain_core.pydantic_v1 import Extra, Field, SecretStr, root_validator
+from langchain_core.utils import convert_to_secret_str, get_from_dict_or_env
+
+from langchain_community.utilities.requests import Requests
+
+
+def _message_role(type: str) -> str:
+    role_mapping = {"ai": "assistant", "human": "user", "chat": "user"}
+
+    if type in role_mapping:
+        return role_mapping[type]
+    else:
+        raise ValueError(f"Unknown type: {type}")
+
+
+def _format_edenai_messages(messages: List[BaseMessage]) -> Dict[str, Any]:
+    system = None
+    formatted_messages = []
+    text = messages[-1].content
+    for i, message in enumerate(messages[:-1]):
+        if message.type == "system":
+            if i != 0:
+                raise ValueError("System message must be at beginning of message list.")
+            system = message.content
+        else:
+            formatted_messages.append(
+                {
+                    "role": _message_role(message.type),
+                    "message": message.content,
+                }
+            )
+    return {
+        "text": text,
+        "previous_history": formatted_messages,
+        "chatbot_global_action": system,
+    }
+
+
+class ChatEdenAI(BaseChatModel):
+    """`EdenAI` chat large language models.
```

Comment:
Could you include a link to EdenAIIAPI docs from developers to reference?

It's unclear from the docs why users should be using EdenAI. The example shown above just shows how to use OpenAI without any context. From my point of view as a user, it feels like I'm just adding network latency and another point of failure if EdenAI is serving just as a proxy. I think it would be good to add a description in the notebook documentation explaining the advantage of using EdenAI and/or link to the relevant explanation on EdenAI's website.

Reply:
I just made some updates to the docs to give a clearer picture of Eden AI and there's now a link to the API reference documentation.